### PR TITLE
feat(workflow): Snapshot AlertRule on update if attached alert exists

### PR DESCRIPTION
**Pull request by iProgramStuff**
When a rule is being updated, we now check if it has any existing incidents attached and snapshot the rule if so. We change the incident's alert rule reference, and mark all incidents associated with the rule as resolved. We'll use the snapshotted data to display on the alert page.

Future PRs will handle other support around this, such as writing to the activity log and frontend checks to inform the user of this behaviour (e.g. something like "You're updating a rule with an active alert associated with it. This will resolve that alert automatically and could cause a new alert to be fired.").

### CODE REVIEW

File: src/sentry/incidents/logic.py [hunk 1]
```diff
@@ -555,6 +555,30 @@ def create_alert_rule(
...
+
+    for trigger in triggers:
+        actions = AlertRuleTriggerAction.objects.filter(alert_rule_trigger=trigger)
+        trigger.id = None
+        trigger.alert_rule = alert_rule
+        trigger.save()
+        for action in actions:
+            action.id = None
+            action.trigger = trigger
+            action.save()
```
Comment (wedamija) [inline src/sentry/incidents/logic.py:579]: Let's put this in a transaction. Even though the caller is already using one, it makes this function safe to call from elsewhere.

File: src/sentry/incidents/logic.py [hunk 2]
```diff
@@ -555,6 +555,30 @@ def create_alert_rule(
...
 
 
+def snapshot_alert_rule(alert_rule):
+    # Creates an archived alert_rule using the same properties as the passed rule
+    # It will also resolve any incidents attached to this rule.
+    triggers = AlertRuleTrigger.objects.filter(alert_rule=alert_rule)
+
+    incidents = Incident.objects.filter(alert_rule=alert_rule)
+
+    alert_rule.id = None
```
Comment (wedamija) [inline src/sentry/incidents/logic.py:565]: I wonder if we can clone the alert rule instance in some way before we modify it. It's a little weird that the snapshot function modifies it. As far as I know it's ok to `deepcopy` an instance

File: tests/sentry/incidents/test_logic.py [hunk 3]
```diff
@@ -852,6 +853,69 @@ def test_update_from_include_all(self):
             sub.project for sub in QuerySubscription.objects.filter(alert_rules=alert_rule)
         ] == [new_project]
 
+    @pytest.mark.chris
```
Comment (wedamija) [inline tests/sentry/incidents/test_logic.py:856]: What's this for?

File: tests/sentry/incidents/test_logic.py [hunk 4]
```diff
@@ -852,6 +853,69 @@ def test_update_from_include_all(self):
...
+
+        # Rule snapshot should have properties of the rule before it was updated.
+        assert rule_snapshot.id != updated_rule.id
+        assert rule_snapshot.name == updated_rule.name
+        assert rule_snapshot.query == "level:error"
+        assert rule_snapshot.time_window == 10
+        assert rule_snapshot.aggregation == QueryAggregations.TOTAL.value
+        assert rule_snapshot.threshold_period == 1
+
+        # Incidents should now be pointing to the rule snapshot.
```
Comment (wedamija) [inline tests/sentry/incidents/test_logic.py:903]: Nit: we could remove duplication a bit here by:
```
for incident in (incident, incident_2):
    # tests
```

File: tests/sentry/incidents/test_logic.py [hunk 5]
```diff
@@ -852,6 +853,69 @@ def test_update_from_include_all(self):
...
+
+        # Incidents should be resolved
+        assert incident.status == IncidentStatus.CLOSED.value
+        assert incident_2.status == IncidentStatus.CLOSED.value
+
+        # Action and trigger counts should double (from 1 to 2)
+        assert AlertRuleTrigger.objects.all().count() == trigger_count * 2
+        assert AlertRuleTriggerAction.objects.all().count() == action_count * 2
+
+        # TODO: Verify actions and triggers have the same properties?
```
Comment (wedamija) [inline tests/sentry/incidents/test_logic.py:917]: Probably worth verifying that they have the same properties, and that they're not associated with the old alert rule.

File: tests/sentry/incidents/test_logic.py (hunk 3 above)
Comment (iProgramStuff) [reply tests/sentry/incidents/test_logic.py:856]: It's leftover from development - it's a simple way to mark tests...then I can simply pytest -m chris to run all the tests marked as `chris`. Was a better alternative than naming tests with _meow :-)

File: tests/sentry/incidents/test_logic.py (hunk 4 above)
Comment (iProgramStuff) [reply tests/sentry/incidents/test_logic.py:903]: Thanks, that's nicer code.

**Review by wedamija — APPROVED** [review]
Looks good to me, thanks for the thorough tests!

**Pull request merged**